File: backend/src/entities/Product.js
import { EntitySchema } from '@mikro-orm/core';

/**
 * @typedef {Object} ProductProperties
 * @property {string} id - UUID primary key
 * @property {string} barcode - Unique barcode
 * @property {string} name - Product name
 * @property {string|null} brand - Brand name
 * @property {string|null} description - Product description
 * @property {string|null} imageUrl - Product image URL
 * @property {string|null} category - Product category
 * @property {string|null} quantity - Product quantity/size
 * @property {Object|null} nutritionData - Nutrition information
 * @property {string} source - Data source (openfoodfacts)
 * @property {Date} createdAt - Creation timestamp
 * @property {Date} updatedAt - Last update timestamp
 * @property {Date|null} lastFetchedAt - Last API fetch timestamp
 */

/**
 * Product Entity Schema
 * @type {EntitySchema<ProductProperties>}
 */
export const ProductSchema = new EntitySchema({
  name: 'Product',
  tableName: 'product',
  properties: {
    id: {
      type: 'uuid',
      primary: true,
      onCreate: () => crypto.randomUUID(),
    },
    barcode: {
      type: 'string',
      unique: true,
    },
    name: {
      type: 'string',
    },
    brand: {
      type: 'string',
      nullable: true,
    },
    description: {
      type: 'text',
      nullable: true,
    },
    imageUrl: {
      type: 'string',
      nullable: true,
    },
    category: {
      type: 'string',
      nullable: true,
    },
    quantity: {
      type: 'string',
      nullable: true,
    },
    nutritionData: {
      type: 'jsonb',
      nullable: true,
    },
    source: {
      type: 'string',
      default: 'openfoodfacts',
    },
    createdAt: {
      type: 'Date',
      onCreate: () => new Date(),
    },
    updatedAt: {
      type: 'Date',
      onCreate: () => new Date(),
      onUpdate: () => new Date(),
    },
    lastFetchedAt: {
      type: 'Date',
      nullable: true,
    },
  },
});

File: mobile/lib/models/product.dart
class Product {
  final String id;
  final String barcode;
  final String name;
  final String? brand;
  final String? description;
  final String? imageUrl;
  final String? category;
  final String? quantity;
  final Map<String, dynamic>? nutritionData;
  final String source;
  final DateTime createdAt;
  final DateTime updatedAt;
  final bool isFavorite;

  Product({
    required this.id,
    required this.barcode,
    required this.name,
    this.brand,
    this.description,
    this.imageUrl,
    this.category,
    this.quantity,
    this.nutritionData,
    this.source = 'openfoodfacts',
    required this.createdAt,
    required this.updatedAt,
    this.isFavorite = false,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'] as String,
      barcode: json['barcode'] as String,
      name: json['name'] as String,
      brand: json['brand'] as String?,
      description: json['description'] as String?,
      imageUrl: json['imageUrl'] as String?,
      category: json['category'] as String?,
      quantity: json['quantity'] as String?,
      nutritionData: json['nutritionData'] as Map<String, dynamic>?,
      source: json['source'] as String? ?? 'openfoodfacts',
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
      isFavorite: json['isFavorite'] as bool? ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'barcode': barcode,
      'name': name,
      'brand': brand,
      'description': description,
      'imageUrl': imageUrl,
      'category': category,
      'quantity': quantity,
      'nutritionData': nutritionData,
      'source': source,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
      'isFavorite': isFavorite,
    };
  }

  Product copyWith({
    String? id,
    String? barcode,
    String? name,
    String? brand,
    String? description,
    String? imageUrl,
    String? category,
    String? quantity,
    Map<String, dynamic>? nutritionData,
    String? source,
    DateTime? createdAt,
    DateTime? updatedAt,
    bool? isFavorite,
  }) {
    return Product(
      id: id ?? this.id,
      barcode: barcode ?? this.barcode,
      name: name ?? this.name,
      brand: brand ?? this.brand,
      description: description ?? this.description,
      imageUrl: imageUrl ?? this.imageUrl,
      category: category ?? this.category,
      quantity: quantity ?? this.quantity,
      nutritionData: nutritionData ?? this.nutritionData,
      source: source ?? this.source,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      isFavorite: isFavorite ?? this.isFavorite,
    );
  }

  String get displayName {
    if (brand != null) {
      return '$brand - $name';
    }
    return name;
  }
}
